Empty PSK Identity

sbutcher-arm opened the pull request:
## Description
This is an update on PR #616 and a fix for #614.

It fixes handling of empty PSK identities.

## Status
**READY**

## Requires Backporting
Yes
Which branch?
`mbedtls-2.16` and `mbedtls-2.7`

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -3048,8 +3049,8 @@ static int ssl_prepare_server_key_exchange( mbedtls_ssl_context *ssl,
 
-        if( ( ret = mbedtls_ecdh_setup( &ssl->handshake->ecdh_ctx,
-                                        (*curve)->grp_id ) ) != 0 )
+        if( ( ret = mbedtls_ecp_group_load( &ssl->handshake->ecdh_ctx.grp,
```
@sbutcher-arm On first sight, this change seems unrelated to the problem this PR sets out to fix. Is it intentional?

sbutcher-arm replied:
That looks like a rather curious mistake in the rebase.

sbutcher-arm commented:
The rebase seems to have introduced some issues. I'll rework the PR and submit it again.

AndrzejKurek commented:
@sbutcher-arm - please re-work this PR and resolve conflicts.

sbutcher-arm commented:
Rebased and reworked (quite radically - this is an *old* PR).

@AndrzejKurek - assuming this passes CI, could I please ask for another review? If it fails, don't bother!

I'll do the backports once it's reviewed and approved.

sbutcher-arm commented:
@hanno-arm - do you want to review again as you're almost still involved? Don't feel obliged! I can find another reviewer if need be.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -1041,9 +1041,11 @@ struct mbedtls_ssl_config
+                                     *   value can be zero to indicate an empty
+                                     *   identity has been set.
+                                     *   If psk_identity is \c NULL its value
+                                     *   must always be 0. */
```
Minor: `\c 0` instead of `0`.

sbutcher-arm replied:
Good catch. Fixed.

hanno-arm commented on library/ssl_cli.c:
```diff
@@ -59,8 +59,7 @@
 {
-    if( conf->psk_identity     == NULL ||
-        conf->psk_identity_len == 0     )
+    if( conf->psk_identity     == NULL )
```
Minor: Indentation is off now that the second line has been removed.

sbutcher-arm replied:
Yup. Fixed.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -4355,7 +4355,9 @@ static int ssl_conf_set_psk_identity( mbedtls_ssl_config *conf,
     }
 
-    conf->psk_identity = mbedtls_calloc( 1, psk_identity_len );
+    /* The PSK Identity can be length zero, but we can't allocate zero bytes,
```
Nitpick: You can call `malloc( 0 )`, but IIRC implementations are allowed to return `NULL` which hence cannot be treated as an error anymore.

sbutcher-arm replied:
Yup, fixed as per Gilles' solution below.

gilles-peskine-arm commented on library/ssl_tls.c:
```diff
@@ -4355,7 +4355,9 @@ static int ssl_conf_set_psk_identity( mbedtls_ssl_config *conf,
+    /* The PSK Identity can be length zero, but we can't allocate zero bytes,
+       so increment, so the buffer is always greater than zero. */
+    conf->psk_identity = mbedtls_calloc( 1, ( psk_identity_len + 1 ) );
     if( conf->psk_identity == NULL )
```
```suggestion
    conf->psk_identity = mbedtls_calloc( 1, psk_identity_len );
    if( psk_identity_len != 0 && conf->psk_identity == NULL )
```

sbutcher-arm replied:
Slightly embarrassed I didn't see the more obvious, simpler and elegant solution you proposed as opposed to my more complex fix - but anyway - I've added that fix.

sbutcher-arm commented:
Rebased, and added in fixes for the issues raised. @gilles-peskine-arm, @hanno-arm - if you could re-review, I'd be grateful.

sbutcher-arm commented:
Failing on the CI when `MBEDTLS_MEMORY_BUFFER_ALLOC_C` is enabled:
```
PSK callback: empty id ................................................. FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-25.log
```
I'll rework it.

tom-daubney-arm commented:
We are now converting older PRs to draft PRs where the following conditions are met: They have not been updated in the last 3 months, and they need more than non-trivial work to complete.
